### ISSUE TITLE
AsyncTest.ParallelRuns unit test is flaky

### ISSUE THREAD

**Issue by valentiniliescu** [refactoring-pipelines/Pipelines#70]
This unit test sometimes fails with these errors:
```
Test method Refactoring.PipelinesAsync.Test.AsyncTest.ParallelRuns threw exception: 
System.AggregateException: One or more errors occurred. (One or more errors occurred. (One or more errors occurred. (Queue empty.))) ---> System.AggregateException: One or more errors occurred. (One or more errors occurred. (Queue empty.)) ---> System.AggregateException: One or more errors occurred. (Queue empty.) ---> System.InvalidOperationException: Queue empty.
    at System.Collections.Generic.Queue`1.ThrowForEmptyQueue()
   at System.Collections.Generic.Queue`1.Dequeue()
   at Refactoring.Pipelines.JoinedPipes`2._SendIfReady() in C:\Users\vilie\Source\Repos\Pipelines\Refactoring.Pipelines\JoinedPipes.cs:line 66
   at Refactoring.Pipelines.JoinedPipes`2.OnMessage2(TInput2 value) in C:\Users\vilie\Source\Repos\Pipelines\Refactoring.Pipelines\JoinedPipes.cs:line 59
   at Refactoring.Pipelines.ForwardingListener`1.OnMessage(T value) in C:\Users\vilie\Source\Repos\Pipelines\Refactoring.Pipelines\ForwardingListener.cs:line 32
```
or
```
Assert.AreEqual failed. Expected:<[1, 2, 3, 4, 5, 6, 7]>. Actual:<[1, 2, 3, 4, 5, 7]>. 
   at Refactoring.PipelinesAsync.Test.AsyncTest.ParallelRuns() in C:\Users\vilie\Source\Repos\Pipelines\Refactoring.Pipelines.Test\AsyncTest.cs:line 39
```

**Comment by valentiniliescu** [comment]
The issue is the async code uses Parallel.ForEach which executes code (usually) multi-threaded. Because of that, the code needs to be thread-safe.
#71 is a draft PR that has some possible fixes.

**Comment by valentiniliescu** [comment]
We should have a discussion about users using async pipelines with code that is not thread-safe.

**Comment by valentiniliescu** [comment]
The issues was fixed in #71 
Added thread-safety issues to follow up in the TODO list
